ros_packages/voice_assistant/voice_assistant/chat.py

```diff
@@ -81,51 +83,104 @@ def chat(self, goal_handle: ServerGoalHandle):
+            if generate_code: description = CODE_DESCRIPTION_PREFIX + description
+            print(description)
             with self.public_voice_client_lock:
                 tokens = public_voice_client.chat_completion(content, description)
 
-            curr_sentence: str = ""
-            prev_sentence: str | None = None
-            sentence_boundary = re.compile(r"[^\d | ^A-Z][\.|!|\?|:]")
+            # regex for indentifying sentences
+            sentence_pattern = re.compile(r"^(?!<pib-program>)(.*?)(([^\d | ^A-Z][\.|!|\?|:])|<pib-program>)", re.DOTALL)
+            # regex-pattern for indentifying visual-code blocks
+            code_visual_pattern = re.compile(r"^<pib-program>(.*?)</pib-program>", re.DOTALL)
+
+            # the text that was currently collected by chaining together tokens
+            # at any given point in time, this string must not contain any leading whitespaces!
+            curr_text: str = ""
+            # previously collected text, that is waiting to be published as feedback
+            prev_text: str | None = None
 
             for token in tokens:
-                # if the goal was cancelled, return immediately
-                if goal_handle.is_cancel_requested:
-                    goal_handle.canceled()
-                    return Chat.Result(rest=curr_sentence)
-                # if a sentence was already found and another token was received, forward the sentence as feedback  
-                if prev_sentence is not None: 
-                    self.executor.create_task(self.create_chat_message, chat_id, prev_sentence, False)
+
+                if prev_text is not None: 
+                    # publish the previously collected text in form of feedback
                     feedback = Chat.Feedback()
-                    feedback.sentence = prev_sentence
+                    feedback.text = prev_text
+                    feedback.text_type = prev_text_type
                     goal_handle.publish_feedback(feedback)
-                    prev_sentence = None
-                # check if the current token marks the end of a sentence
-                if sentence_boundary.search(curr_sentence):
-                    prev_sentence = curr_sentence.strip()
-                    curr_sentence = ""
-                curr_sentence += token
+                    prev_text = None
+                    prev_text_type = None
+
+                # add token to current text; remove leading white-spaces, if current-text is empty
+                curr_text = curr_text + (token if len(curr_text) > 0 else token.lstrip())
+
+                while True: # loop until current-text was not stripped during current iteration
+
+                    # if the goal was cancelled, return immediately
+                    if goal_handle.is_cancel_requested:
+                        goal_handle.canceled()
+                        return Chat.Result() # TODO
```

Comment:
Is the ToDo part of this story? If not, add a description for the future

Reply:
removed